src/frequenz/sdk/timeseries/periodic_feature_extractor.py

```diff
@@ -0,0 +1,392 @@
+
+        self._period = int(period.total_seconds() / self._sampling_period)
+        """Distance between two succeeding intervals in samples."""
+
+        log.debug("len(moving_window): %i", len(self._moving_window))
+        log.debug("period: %i", self._period)
+
+        if not len(self._moving_window) % self._period == 0:
+            raise ValueError(
+                "The MovingWindow size is not a integer multiple of the period."
+            )
+
+        if self._period - period.total_seconds() / self._sampling_period != 0:
+            log.warning(
+                "The period is not a multiple of the sampling period. "
+                "This might result in unexpected behaviour."
+            )
+
+        log.debug("Initializing PeriodicFeatureExtractor!")
+        log.debug(
+            "Period between two succeeding intervals (in samples): %i",
+            self._period,
+        )
+
+    def _timestamp_to_rel_index(self, timestamp: datetime) -> int:
+        """
+        Get the index of a timestamp relative to the oldest sample in the MovingWindow.
+
+        In other word consider an integer axis where the zero is defined as the
+        oldest element in the MovingWindow. This function returns the index of
+        the given timestamp an this axis.
+
+        Args:
+            timestamp: A timestamp that we want to shift into the window.
+
+        Returns:
+            The index of the timestamp shifted into the MovingWindow.
+        """
+        # align timestamp to the sampling period
+        timestamp = self._buffer.normalize_timestamp(timestamp)
+
+        # distance between the input ts and the ts of oldest know samples (in samples)
+        dist_to_oldest = int(
+            abs((timestamp - self._moving_window.time_bound_oldest).total_seconds())
+            / self._sampling_period
+        )
+
+        log.debug("Shifting ts: %s", timestamp)
+        log.debug(
+            "Oldest timestamp in buffer: %s", self._moving_window.time_bound_oldest
+        )
+        log.debug("Distance to the oldest sample: %i", dist_to_oldest)
+
+        return dist_to_oldest
+
+    def _num_windows(
+        self, window: NDArray[np.float_] | MovingWindow, window_size: int
+    ) -> int:
+        """
+        Get the number of windows that are fully contained in the MovingWindow.
+
+        This method calculates how often a given window, defined by it size, is
+        fully contained in the MovingWindow or any numpy ndarray given the
+        period between two window neighbors.
+
+        Args:
+            window: The buffer that is used for the average calculation.
+            window_size: The size of the window in samples.
+
+        Returns:
+            The number of windows that are fully contained in the MovingWindow.
+        """
+        num_windows = math.floor(len(window) / self._period)
+        if len(window) - num_windows * self._period >= window_size:
```

Comment:
When can this happen?

Reply:
This can't happen any longer with the current restrictions. This function was actually meant to work in a more generic situation. I'll comment that code out and leave it was a comment for future reference.

Reply:
I moved that function to the benchmarks as at hasn't been used in the code at all.